decisions/2024-02-28_file_sync.md

```diff
@@ -0,0 +1,91 @@
+# File Sync
+
+- _Date_: 2024-02-28
+- _Deciders_: James Brunskill, Mark Prins, Andrei Evguenov, Chris Petty
+- _Status_: DECIDED (almost)
+- _Outcome_: Option 1
+
+## Context
+
+Our cold chain equipment management features require a mechanism to upload and download documents, and images related to assets.
+
+If these were to go via our standard sync process, large files would potentially block the sync process, slow down deployments to sites, etc.
+
+All options assume we'll use the open-mSupply Central Server Sync process recently established.
+This means that file sync will only work for open-mSupply Installations.
+
+## Options
+
+### Option 1 - Sync File IDs via std process, and download/upload files in a separate process
+
+Since we already have a process which allows for syncing data, controlling which records go to which sites, etc, we can reuse this process, to send a list of file ID to each site. Once the file ID is know the server can kick off a background process to download the file to the site via HTTP. If a file hasn't been downloaded yet, but is requested by the client, the server could download the file immediately and cache the result. (This could have some obvious challenges with files that are already halfway downloaded, connection issues, etc.)
+
+One key advantage of this approach is that we should be able to store partially downloaded (and uploaded) files and continue from where they're left off rather than having the whole sync request fail.
+
+We can also put in additional logic such as suspending file sync while the main sync is in progress, rate limiting, etc.
+
+It is assumed that we won't do anything special to compress files in MVP.
+A similar process can be implemented for uploading files using [TUS](tus.io)
+
+![File Sync](./media/file_sync.png)
+
+_Pros:_
+
+- Keeps the main sync process simple/unchanged
+- Allows for targeted logic for file syncing (e.g rate limiting in future)
+- Allows for files to be downloaded where they're needed rather than needing to sync everywhere (via std sync process)
+
+_Cons:_
+
+- We could end up writing quite a bit of code to implement this and may have unexpected challenges with the implementation
+- Other sync protocols might provide benefits e.g. rsync for example uses a delta transfer algorithm which only sends the differences between files, rather than the whole file.
+
+
+### Option 2 - Base64 encode files and use std sync
+
+Any files needed to be synced could be base64 encoded and sent as part of the normal sync process. This would be a simiplish process, but would mean that sync would be blocked until all files are fully downloaded. Base64 encoding would also increase the size of the file by 33% (or more) which could be a problem for large files.
```

Comment:
```suggestion
Any files needed to be synced could be base64 encoded and sent as part of the normal sync process. This would be a simplish process, but would mean that sync would be blocked until all files are fully downloaded. Base64 encoding would also increase the size of the file by 33% (or more) which could be a problem for large files.
```